File: beautiful.java
// Time Complexity : O(n!) 
// Space Complexity : O(k) ==> size of number
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no


// Your code here along with comments explaining your approach


class Solution {
    int result=0;
    public int countArrangement(int n) {
        
        boolean[] visited=new boolean[n+1];
        
        helper(visited,n,1);
        return result;
    }
    
    private void helper(boolean[] visited, int size, int index){
        
        //base case
        if(index > size){
             result++;
          //  System.out.println();
           
            
        }
           
        
        
        // logic 
        for(int i=1;i<=size;i++){
            
            if(!visited[i] && (i%index==0 || index%i==0)){
                //action
                visited[i]=true;
                
                //recurse
                helper(visited,size, index+1);
                
                // backtrack
                visited[i]=false;
                
            }
            
        }
        
        
    }
}

File: Beautiful.cpp
// TC: O(N!) where N is valid palindromes
// SC: O(h) where h is the height of stack

class Solution {
public:
    int count=0;
    int countArrangement(int n) {
        if(n==1)
            return 1;
        vector<int> nums;
        for(int i=1; i<=n; i++)
            nums.push_back(i);
        helper(nums, 0);
        return count;
    }

    //helper
    void helper(vector<int> nums, int index){
        //base condition
        if(index == nums.size()){
            count++;
            return;
        }
        //iterative recursion
        for(int i= index; i< nums.size(); i++){
            // check valid
            if(nums[i] % (index+1) == 0 || (index+1) % nums[i] ==0 ){
                //action
                swap(nums[index], nums[i]);
                //recurse
                helper(nums, index+1);
                //backtrack
                swap(nums[i], nums[index]);
            }
        }
    }
};